File: LeetCode/0740. Delete and Earn/solution.py
"""
47 / 47 test cases passed.
Runtime: 36 ms
Memory Usage: 15 MB
"""
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        freq = [0] * (max(nums) + 1)
        for num in nums:
            freq[num] += num
        last, curr = 0, 0
        for cnt in freq:
            last, curr = curr, max(curr, last + cnt)
        return curr

File: LeetCode/0740. Delete and Earn/solution.cpp
/**
 * 47 / 47 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 8.9 MB 
 */
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int max_ = *max_element(nums.begin(), nums.end());
        vector<int> freq(max_ + 1);
        for (auto& num: nums) {
            freq[num] += num;
        }
        int last = 0, curr = 0;
        for (auto& cnt: freq) {
            int tmp = curr;
            curr = max(curr, last + cnt);
            last = tmp;
        }
        return curr;
    }
};
